src/apps/vpn/connectionmanager.cpp

```diff
@@ -0,0 +1,1000 @@
+          &ConnectionManager::maybeEnableDisconnectInConfirming);
+
+  connect(&m_ping_canary, &PingHelper::pingSentAndReceived, this, [this]() {
+    m_ping_canary.stop();
+    m_ping_received = true;
+    logger.info() << "Canary Ping Succeeded";
+  });
+
+  connect(SettingsHolder::instance(), &SettingsHolder::transactionBegan, this,
+          [this]() {
+            m_connectedBeforeTransaction =
+                m_state == ConnectionManager::StateOn;
+          });
+
+  connect(SettingsHolder::instance(),
+          &SettingsHolder::transactionAboutToRollBack, this, [this]() {
+            if (!m_connectedBeforeTransaction)
+              MozillaVPN::instance()->deactivate();
+          });
+
+  connect(SettingsHolder::instance(), &SettingsHolder::transactionRolledBack,
+          this, [this]() {
+            if (m_connectedBeforeTransaction)
+              MozillaVPN::instance()->activate();
+          });
+
+  MozillaVPN* vpn = MozillaVPN::instance();
+
+  const Device* device = vpn->deviceModel()->currentDevice(vpn->keys());
+  m_impl->initialize(device, vpn->keys());
+
+  connect(SettingsHolder::instance(), &SettingsHolder::serverDataChanged, this,
+          &ConnectionManager::serverDataChanged);
+
+  connect(LogHandler::instance(), &LogHandler::cleanupLogsNeeded, this,
+          &ConnectionManager::cleanupBackendLogs);
+
+  connect(this, &ConnectionManager::readyToServerUnavailable,
+          Tutorial::instance(), &Tutorial::stop);
+}
+
+void ConnectionManager::implInitialized(bool status, bool a_connected,
+                                        const QDateTime& connectionDate) {
+  logger.debug() << "Controller initialized with status:" << status
+                 << "connected:" << a_connected
+                 << "connectionDate:" << connectionDate.toString();
+
+  Q_ASSERT(m_state == ConnectionManager::StateInitializing);
+
+  if (!status) {
+    REPORTERROR(ErrorHandler::ControllerError, "controller");
+    setState(StateOff);
+    return;
+  }
+
+  if (processNextStep()) {
+    setState(StateOff);
+    return;
+  }
+
+  setState(a_connected ? StateOn : StateOff);
+
+  // If we are connected already at startup time, we can trigger the connection
+  // sequence of tasks.
+  if (a_connected) {
+    m_connectedTimeInUTC = connectionDate.isValid()
+                               ? connectionDate.toUTC()
+                               : QDateTime::currentDateTimeUtc();
+    emit timeChanged();
+    m_timer.start(TIMER_MSEC);
+  }
+}
+
+qint64 ConnectionManager::time() const {
+  if (m_connectedTimeInUTC.isValid()) {
+    return m_connectedTimeInUTC.secsTo(QDateTime::currentDateTimeUtc());
+  }
+  return 0;
+}
+
+void ConnectionManager::timerTimeout() {
+  Q_ASSERT(m_state == StateOn);
+  emit timeChanged();
+}
+
+void ConnectionManager::quit() {
+  logger.debug() << "Quitting";
+
+  if (m_state == StateInitializing || m_state == StateOff) {
+    emit readyToQuit();
+    return;
```

Comment:
Is it OK that m_nextStep is not set to Quit here? (This is old code, so just checking.)

Reply:
m_nextStep is set to quit below this if statement. Are you asking if it should also be set to Quit within this?

Reply:
Yes, because the app's slot calls App::quit on the readyToQuit signal, and ConnectionManager's m_state is not updated to Quit because of the early return on line 230.

A change may not be necessary because the app is t, but updating the state may help to help future debugging during app shutdown. Feel free to ignore this comment.

Reply:
@vinocher Your logic makes a lot of sense to me so I went ahead and added your suggestion to the if statement above. If 'll double check with Naomi to make sure we're not missing something.